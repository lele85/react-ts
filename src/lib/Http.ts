import Axios from "axios";

function get<T> (uri: string) : Promise<T> {
    return Axios.get(uri).then((res) => {
        const response : T = res.data;
        return response;
    });
};

export default {
    get: get
};