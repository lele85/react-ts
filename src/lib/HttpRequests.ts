import { set, get } from "lodash";
import { AbortTokens } from "../actions/ShowActions";
import Axios from "axios";

const IDENTITY = (x: any) => x;

const requests = {};

const getHttpRequest = (token: AbortTokens) => {
    return get(requests, `${token}`, { cancel: IDENTITY, token: null });
};

const setHttpRequest = (token: AbortTokens) => {
    if (token) {
        const source = Axios.CancelToken.source();
        set(requests, `${token}`, source);
    }
};

const abortHttpRequest = (token: AbortTokens) => {
    getHttpRequest(token).cancel();
};

export default {
    getHttpRequest,
    setHttpRequest,
    abortHttpRequest
};
