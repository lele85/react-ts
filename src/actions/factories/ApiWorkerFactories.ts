import { bindThunkAction } from "typescript-fsa-redux-thunk";
import Http from "../../lib/Http";
import { AsyncActionCreators } from "typescript-fsa";

export function fetchApiWorkerFactory<TParams, TResponse, TError> (asyncCreator : AsyncActionCreators<TParams,TResponse, TError>, path: string) {
    return bindThunkAction(
        asyncCreator,
        async (params, dispatch) => {
            const result = await Http.get<TResponse>(path);
            return result;
        }
    )
}