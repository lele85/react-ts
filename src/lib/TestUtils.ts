import Http from "./Http";
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';

export function mockGetSuccess<T>(result : T) {
    Http.get = jest.fn<T>(() => {
        return Promise.resolve(result);
    });
};

export function mockGetError<T>(error : T) {
    Http.get = jest.fn<T>(() => {
        return Promise.reject(error);
    });
};

export function createMockStore<T>(state: T) {
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);
    return mockStore(state);
}

export function assertEquals<T> (result : T, expected: T) {
    return expect(result).toEqual(expected);
}