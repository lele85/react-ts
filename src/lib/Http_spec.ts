import Http, { buildUri } from "./Http";
import { AbortTokens } from "../actions/ShowActions";
import Axios from "axios";
import HttpRequests from "./HttpRequests";
describe("Http", () => {
    describe("get", () => {
        const mockModel = {
            id: 12,
            title: "Show Title"
        };
        const mockParams = {
            id: 12,
            exports: "all"
        };
        const mockCancel = {
            cancel: () => {},
            token: "123"
        };

        beforeEach(() => {
            jest.spyOn(HttpRequests, "getHttpRequest").mockReturnValue(mockCancel);
            jest.spyOn(HttpRequests, "setHttpRequest");
            jest.spyOn(HttpRequests, "abortHttpRequest");
        });

        it("should cancel the pending request with the same token, set the current one and resolve succesfully", async () => {
            jest.spyOn(Axios, "get").mockReturnValue(Promise.resolve({ data: mockModel }));
            const result = await Http.get<typeof mockModel, typeof mockParams>(
                "/api/shows/:id",
                mockParams,
                AbortTokens.SHOW
            );
            expect(result).toEqual({ canceled: false, model: mockModel });
            expect(HttpRequests.getHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
            expect(HttpRequests.setHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
            expect(HttpRequests.abortHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
            expect(Axios.get).toHaveBeenCalledWith("/api/shows/12", { cancelToken: "123", exports: "all" });
        });

        it("should cancel the pending request with the same token, set the current one and reject in case of error", async () => {
            jest.spyOn(Axios, "get").mockReturnValue(Promise.reject({ error: new Error("message") }));
            try {
                await Http.get<typeof mockModel, typeof mockParams>("/api/shows/:id", mockParams, AbortTokens.SHOW);
            } catch (e) {
                expect(HttpRequests.getHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
                expect(HttpRequests.setHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
                expect(HttpRequests.abortHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
                expect(Axios.get).toHaveBeenCalledWith("/api/shows/12", { cancelToken: "123", exports: "all" });
            }
        });

        it("should cancel the pending request with the same token, set the current one and resolve with a canceled response in case of CancelationError", async () => {
            jest.spyOn(Axios, "get").mockReturnValue(Promise.reject(new Axios.Cancel("message")));
            const result = await Http.get<typeof mockModel, typeof mockParams>(
                "/api/shows/:id",
                mockParams,
                AbortTokens.SHOW
            );
            expect(result).toEqual({ canceled: true, model: null });
            expect(HttpRequests.getHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
            expect(HttpRequests.setHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
            expect(HttpRequests.abortHttpRequest).toHaveBeenCalledWith(AbortTokens.SHOW);
            expect(Axios.get).toHaveBeenCalledWith("/api/shows/12", { cancelToken: "123", exports: "all" });
        });
    });

    describe("buildUri", () => {
        it("should return the same uri if no placeholder and no params", () => {
            expect(buildUri("/foo/bar", {})).toEqual({ uri: "/foo/bar", params: {} });
        });

        it("should replace the placholder if there's a match and remove the bound param", () => {
            expect(buildUri("/foo/bar/:id", { id: 12 })).toEqual({
                uri: "/foo/bar/12",
                params: {}
            });
        });

        it("should replace the placholder if there's a match and remove the bound param", () => {
            expect(buildUri("/foo/bar/:id", { id: 12, q: "search" })).toEqual({
                uri: "/foo/bar/12",
                params: { q: "search" }
            });
        });

        it("should replace the placeholedet with a match for value 0", () => {
            expect(buildUri("/foo/bar/:id", { id: 0, q: "search" })).toEqual({
                uri: "/foo/bar/0",
                params: { q: "search" }
            });
        });
    });
});
