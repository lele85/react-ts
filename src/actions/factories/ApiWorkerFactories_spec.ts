import { actionCreatorFactory } from 'typescript-fsa';
import Http from '../../lib/Http';
import { fetchApiWorkerFactory } from './ApiWorkerFactories';
import { createMockStore, mockGetSuccess, mockGetError } from '../../lib/TestUtils';

describe("fetchApiWorkerFactory", () => {

    type State = {};
    type Params = { query: string };
    type Response = { value: number };
    type Error = { message: string };

    it("should dispatch started then done in case of success", async () => { 
        const store = createMockStore<State>({});
        mockGetSuccess<Response>({value:12});
        const actionCreator = actionCreatorFactory("@@test");
        const fetchTest = actionCreator.async<Params,Response,Error>('FETCH');
        const fetchTestWorker = fetchApiWorkerFactory<Params,Response,Error>(fetchTest, "/api/test");
        await store.dispatch(fetchTestWorker({query: "123"}));
        expect(store.getActions()).toMatchSnapshot();
    });

    it("should dispatch started then fail in case of error", async () => {
        const store = createMockStore<State>({});
        mockGetError<Error>({message: "Error!"});
        const actionCreator = actionCreatorFactory("@@test");
        const fetchTest = actionCreator.async<Params,Response,Error>('FETCH');
        const fetchTestWorker = fetchApiWorkerFactory<Params,Response,Error>(fetchTest, "/api/test");
        await store.dispatch(fetchTestWorker({query: "123"}));
        expect(store.getActions()).toMatchSnapshot();
    });
});