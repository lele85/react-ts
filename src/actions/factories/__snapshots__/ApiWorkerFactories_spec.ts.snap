// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchApiWorkerFactory should dispatch started then done in case of success 1`] = `
Array [
  Object {
    "payload": Object {
      "query": "123",
    },
    "type": "@@test/FETCH_STARTED",
  },
  Object {
    "payload": Object {
      "params": Object {
        "query": "123",
      },
      "result": Object {
        "value": 12,
      },
    },
    "type": "@@test/FETCH_DONE",
  },
]
`;

exports[`fetchApiWorkerFactory should dispatch started then fail in case of error 1`] = `
Array [
  Object {
    "payload": Object {
      "query": "123",
    },
    "type": "@@test/FETCH_STARTED",
  },
  Object {
    "error": true,
    "payload": Object {
      "error": Object {
        "message": "Error!",
      },
      "params": Object {
        "query": "123",
      },
    },
    "type": "@@test/FETCH_FAILED",
  },
]
`;
